File: common/src/main/java/com/cobblemon/mod/common/cobblemonstructures/CobblemonStructures.java
package com.cobblemon.mod.common.cobblemonstructures;

import com.cobblemon.mod.common.mixin.StructurePoolAccessor;
import com.mojang.datafixers.util.Pair;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.registry.entry.RegistryEntry;
import net.minecraft.server.MinecraftServer;
import net.minecraft.structure.pool.LegacySinglePoolElement;
import net.minecraft.structure.pool.SinglePoolElement;
import net.minecraft.structure.pool.StructurePool;
import net.minecraft.structure.pool.StructurePoolElement;
import net.minecraft.structure.processor.StructureProcessorList;
import net.minecraft.structure.processor.StructureProcessorLists;
import net.minecraft.util.Identifier;

import java.util.ArrayList;
import java.util.List;

public class CobblemonStructures {

    private static final RegistryKey<StructureProcessorList> EMPTY_PROCESSOR_LIST_KEY = RegistryKey.of(RegistryKeys.PROCESSOR_LIST, new Identifier("minecraft", "empty"));
    private static final Integer pokecenterWeight = 75;
    private static final Integer longPathWeight = 10;

    public static void registerJigsaws(MinecraftServer server) {
        Registry<StructurePool> templatePoolRegistry = server.getRegistryManager().get(RegistryKeys.TEMPLATE_POOL);
        Registry<StructureProcessorList> processorListRegistry = server.getRegistryManager().get(RegistryKeys.PROCESSOR_LIST);

        addPokecenters(templatePoolRegistry, processorListRegistry);
        addLongPaths(templatePoolRegistry, processorListRegistry);
    }

    private static void addPokecenters(Registry<StructurePool> templatePoolRegistry, Registry<StructureProcessorList> processorListRegistry) {
        Identifier plainsHousesPoolLocation = new Identifier("minecraft:village/plains/houses");
        Identifier desertHousesPoolLocation = new Identifier("minecraft:village/desert/houses");
        Identifier savannaHousesPoolLocation = new Identifier("minecraft:village/savanna/houses");
        Identifier snowyHousesPoolLocation = new Identifier("minecraft:village/snowy/houses");
        Identifier taigaHousesPoolLocation = new Identifier("minecraft:village/taiga/houses");

        addBuildingToPool(templatePoolRegistry, processorListRegistry, plainsHousesPoolLocation, CobblemonStructureIDs.PLAINS_POKECENTER, pokecenterWeight, StructurePool.Projection.RIGID, EMPTY_PROCESSOR_LIST_KEY);
        addBuildingToPool(templatePoolRegistry, processorListRegistry, desertHousesPoolLocation, CobblemonStructureIDs.DESERT_POKECENTER, pokecenterWeight, StructurePool.Projection.RIGID, EMPTY_PROCESSOR_LIST_KEY);
        addBuildingToPool(templatePoolRegistry, processorListRegistry, savannaHousesPoolLocation, CobblemonStructureIDs.SAVANNA_POKECENTER, pokecenterWeight, StructurePool.Projection.RIGID, EMPTY_PROCESSOR_LIST_KEY);
        addBuildingToPool(templatePoolRegistry, processorListRegistry, snowyHousesPoolLocation, CobblemonStructureIDs.SNOWY_POKECENTER, pokecenterWeight, StructurePool.Projection.RIGID, EMPTY_PROCESSOR_LIST_KEY);
        addBuildingToPool(templatePoolRegistry, processorListRegistry, taigaHousesPoolLocation, CobblemonStructureIDs.TAIGA_POKECENTER, pokecenterWeight, StructurePool.Projection.RIGID, StructureProcessorLists.MOSSIFY_10_PERCENT);
    }
    private static void addLongPaths(Registry<StructurePool> templatePoolRegistry, Registry<StructureProcessorList> processorListRegistry) {
        Identifier plainsStreetsPoolLocation = new Identifier("minecraft:village/plains/streets");
        Identifier desertStreetsPoolLocation = new Identifier("minecraft:village/desert/streets");
        Identifier savannaStreetsPoolLocation = new Identifier("minecraft:village/savanna/streets");
        Identifier snowyStreetsPoolLocation = new Identifier("minecraft:village/snowy/streets");
        Identifier taigaStreetsPoolLocation = new Identifier("minecraft:village/taiga/streets");

        addLegacyBuildingToPool(templatePoolRegistry, processorListRegistry, plainsStreetsPoolLocation, CobblemonStructureIDs.PLAINS_LONG_PATH, longPathWeight, StructurePool.Projection.TERRAIN_MATCHING, StructureProcessorLists.STREET_PLAINS);
        addLegacyBuildingToPool(templatePoolRegistry, processorListRegistry, desertStreetsPoolLocation, CobblemonStructureIDs.DESERT_LONG_PATH, longPathWeight, StructurePool.Projection.TERRAIN_MATCHING, EMPTY_PROCESSOR_LIST_KEY);
        addLegacyBuildingToPool(templatePoolRegistry, processorListRegistry, savannaStreetsPoolLocation, CobblemonStructureIDs.SAVANNA_LONG_PATH, longPathWeight, StructurePool.Projection.TERRAIN_MATCHING, StructureProcessorLists.STREET_SAVANNA);
        addLegacyBuildingToPool(templatePoolRegistry, processorListRegistry, snowyStreetsPoolLocation, CobblemonStructureIDs.SNOWY_LONG_PATH, longPathWeight, StructurePool.Projection.TERRAIN_MATCHING, StructureProcessorLists.STREET_SNOWY_OR_TAIGA);
        addLegacyBuildingToPool(templatePoolRegistry, processorListRegistry, taigaStreetsPoolLocation, CobblemonStructureIDs.TAIGA_LONG_PATH, longPathWeight, StructurePool.Projection.TERRAIN_MATCHING, StructureProcessorLists.STREET_SNOWY_OR_TAIGA);
    }

    public static void addLegacyBuildingToPool(Registry<StructurePool> templatePoolRegistry, Registry<StructureProcessorList> processorListRegistry, Identifier poolRL, String nbtPieceRL, int weight, StructurePool.Projection projection, RegistryKey<StructureProcessorList> processorListKey) {
        addBuildingToPool(templatePoolRegistry, processorListRegistry, poolRL, nbtPieceRL, weight, projection, processorListKey, true);
    }

    public static void addBuildingToPool(Registry<StructurePool> templatePoolRegistry, Registry<StructureProcessorList> processorListRegistry, Identifier poolRL, String nbtPieceRL, int weight, StructurePool.Projection projection, RegistryKey<StructureProcessorList> processorListKey) {
        addBuildingToPool(templatePoolRegistry, processorListRegistry, poolRL, nbtPieceRL, weight, projection, processorListKey, false);
    }

    public static void addBuildingToPool(Registry<StructurePool> templatePoolRegistry, Registry<StructureProcessorList> processorListRegistry, Identifier poolRL, String nbtPieceRL, int weight, StructurePool.Projection projection, RegistryKey<StructureProcessorList> processorListKey, boolean shouldUseLegacySingePoolElement) {
        if (processorListRegistry.getEntry(processorListKey).isEmpty()) return;
        RegistryEntry.Reference<StructureProcessorList> processorList = processorListRegistry.getEntry(processorListKey).get();

        StructurePool pool = templatePoolRegistry.get(poolRL);
        if (pool == null) return;

        SinglePoolElement piece;
        if (shouldUseLegacySingePoolElement) {
            piece = LegacySinglePoolElement.ofProcessedLegacySingle(nbtPieceRL, processorList).apply(projection);
        } else {
            piece = SinglePoolElement.ofProcessedSingle(nbtPieceRL, processorList).apply(projection);
        }

        for (int i = 0; i < weight; i++) {
            ((StructurePoolAccessor) pool).getElements().add(piece);
        }

        List<Pair<StructurePoolElement, Integer>> listOfPieceEntries = new ArrayList<>(((StructurePoolAccessor) pool).getElementCounts());
        listOfPieceEntries.add(new Pair<>(piece, weight));
        ((StructurePoolAccessor) pool).setElementCounts(listOfPieceEntries);
    }
}

File: common/src/main/kotlin/com/cobblemon/mod/common/world/CobblemonStructures.kt
/*
 * Copyright (C) 2023 Cobblemon Contributors
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.cobblemon.mod.common.world

import com.cobblemon.mod.common.mixin.StructurePoolAccessor
import com.cobblemon.mod.common.world.structureprocessors.CobblemonStructureProcessorLists
import com.mojang.datafixers.util.Pair
import net.minecraft.core.Registry
import net.minecraft.core.registries.Registries
import net.minecraft.data.worldgen.ProcessorLists
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.MinecraftServer
import net.minecraft.world.level.levelgen.structure.pools.LegacySinglePoolElement
import net.minecraft.world.level.levelgen.structure.pools.SinglePoolElement
import net.minecraft.world.level.levelgen.structure.pools.StructureTemplatePool
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureProcessorList

object CobblemonStructures {
    private val EMPTY_PROCESSOR_LIST_KEY = ResourceKey.create(Registries.PROCESSOR_LIST, ResourceLocation.fromNamespaceAndPath("minecraft", "empty"))
    private const val pokecenterWeight = 35
    private const val berryFarmWeight = 1
    private const val longPathWeight = 10

    val plainsHousesPoolLocation = ResourceLocation.fromNamespaceAndPath("minecraft", "village/plains/houses")
    val desertHousesPoolLocation = ResourceLocation.fromNamespaceAndPath("minecraft", "village/desert/houses")
    val savannaHousesPoolLocation = ResourceLocation.fromNamespaceAndPath("minecraft", "village/savanna/houses")
    val snowyHousesPoolLocation = ResourceLocation.fromNamespaceAndPath("minecraft", "village/snowy/houses")
    val taigaHousesPoolLocation = ResourceLocation.fromNamespaceAndPath("minecraft", "village/taiga/houses")

    fun registerJigsaws(server: MinecraftServer) {
        val templatePoolRegistry = server.registryAccess().registryOrThrow(Registries.TEMPLATE_POOL)
        val processorListRegistry = server.registryAccess().registryOrThrow(Registries.PROCESSOR_LIST)

        //No pokecenters yet
        //addPokecenters(templatePoolRegistry, processorListRegistry);
        //addLongPaths(templatePoolRegistry, processorListRegistry);
        addBerryFarms(templatePoolRegistry, processorListRegistry)
    }

    fun addBerryFarms(
        templatePoolRegistry: Registry<StructureTemplatePool>,
        processorListRegistry: Registry<StructureProcessorList>
    ) {
        val cropToBerryProcessorList = CobblemonStructureProcessorLists.CROP_TO_BERRY//ResourceKey.create(ResourceKeys.PROCESSOR_LIST, CobblemonProcessorTypes.RANDOM_POOLED_STATES_KEY)

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            plainsHousesPoolLocation,
            CobblemonStructureIDs.PLAINS_BERRY_SMALL,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            plainsHousesPoolLocation,
            CobblemonStructureIDs.PLAINS_BERRY_LARGE,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            desertHousesPoolLocation,
            CobblemonStructureIDs.DESERT_BERRY_SMALL,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            desertHousesPoolLocation,
            CobblemonStructureIDs.DESERT_BERRY_LARGE,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            savannaHousesPoolLocation,
            CobblemonStructureIDs.SAVANNA_BERRY_SMALL,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            savannaHousesPoolLocation,
            CobblemonStructureIDs.SAVANNA_BERRY_LARGE,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            taigaHousesPoolLocation,
            CobblemonStructureIDs.TAIGA_BERRY_SMALL,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            taigaHousesPoolLocation,
            CobblemonStructureIDs.TAIGA_BERRY_LARGE,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            snowyHousesPoolLocation,
            CobblemonStructureIDs.SNOWY_BERRY_SMALL,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )

        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            snowyHousesPoolLocation,
            CobblemonStructureIDs.SNOWY_BERRY_LARGE,
            berryFarmWeight,
            StructureTemplatePool.Projection.RIGID,
            cropToBerryProcessorList
        )
    }

    private fun addPokecenters(
        templatePoolRegistry: Registry<StructureTemplatePool>,
        processorListRegistry: Registry<StructureProcessorList>
    ) {
        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            plainsHousesPoolLocation,
            CobblemonStructureIDs.PLAINS_POKECENTER,
            pokecenterWeight,
            StructureTemplatePool.Projection.RIGID,
            EMPTY_PROCESSOR_LIST_KEY
        )
        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            desertHousesPoolLocation,
            CobblemonStructureIDs.DESERT_POKECENTER,
            pokecenterWeight,
            StructureTemplatePool.Projection.RIGID,
            EMPTY_PROCESSOR_LIST_KEY
        )
        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            savannaHousesPoolLocation,
            CobblemonStructureIDs.SAVANNA_POKECENTER,
            pokecenterWeight,
            StructureTemplatePool.Projection.RIGID,
            EMPTY_PROCESSOR_LIST_KEY
        )
        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            snowyHousesPoolLocation,
            CobblemonStructureIDs.SNOWY_POKECENTER,
            pokecenterWeight,
            StructureTemplatePool.Projection.RIGID,
            EMPTY_PROCESSOR_LIST_KEY
        )
        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            taigaHousesPoolLocation,
            CobblemonStructureIDs.TAIGA_POKECENTER,
            pokecenterWeight,
            StructureTemplatePool.Projection.RIGID,
            ProcessorLists.MOSSIFY_10_PERCENT
        )
    }

    private fun addLongPaths(
        templatePoolRegistry: Registry<StructureTemplatePool>,
        processorListRegistry: Registry<StructureProcessorList>
    ) {
        val plainsStreetsPoolLocation = ResourceLocation.parse("minecraft:village/plains/streets")
        val desertStreetsPoolLocation = ResourceLocation.parse("minecraft:village/desert/streets")
        val savannaStreetsPoolLocation = ResourceLocation.parse("minecraft:village/savanna/streets")
        val snowyStreetsPoolLocation = ResourceLocation.parse("minecraft:village/snowy/streets")
        val taigaStreetsPoolLocation = ResourceLocation.parse("minecraft:village/taiga/streets")
        addLegacyBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            plainsStreetsPoolLocation,
            CobblemonStructureIDs.PLAINS_LONG_PATH,
            longPathWeight,
            StructureTemplatePool.Projection.TERRAIN_MATCHING,
            ProcessorLists.STREET_PLAINS
        )
        addLegacyBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            desertStreetsPoolLocation,
            CobblemonStructureIDs.DESERT_LONG_PATH,
            longPathWeight,
            StructureTemplatePool.Projection.TERRAIN_MATCHING,
            EMPTY_PROCESSOR_LIST_KEY
        )
        addLegacyBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            savannaStreetsPoolLocation,
            CobblemonStructureIDs.SAVANNA_LONG_PATH,
            longPathWeight,
            StructureTemplatePool.Projection.TERRAIN_MATCHING,
            ProcessorLists.STREET_SAVANNA
        )
        addLegacyBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            snowyStreetsPoolLocation,
            CobblemonStructureIDs.SNOWY_LONG_PATH,
            longPathWeight,
            StructureTemplatePool.Projection.TERRAIN_MATCHING,
            ProcessorLists.STREET_SNOWY_OR_TAIGA
        )
        addLegacyBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            taigaStreetsPoolLocation,
            CobblemonStructureIDs.TAIGA_LONG_PATH,
            longPathWeight,
            StructureTemplatePool.Projection.TERRAIN_MATCHING,
            ProcessorLists.STREET_SNOWY_OR_TAIGA
        )
    }

    fun addLegacyBuildingToPool(
        templatePoolRegistry: Registry<StructureTemplatePool>,
        processorListRegistry: Registry<StructureProcessorList>,
        poolRL: ResourceLocation,
        nbtPieceRL: ResourceLocation,
        weight: Int,
        projection: StructureTemplatePool.Projection,
        processorListKey: ResourceKey<StructureProcessorList>
    ) {
        addBuildingToPool(
            templatePoolRegistry,
            processorListRegistry,
            poolRL,
            nbtPieceRL,
            weight,
            projection,
            processorListKey,
            true
        )
    }

    @JvmOverloads
    fun addBuildingToPool(
        templatePoolRegistry: Registry<StructureTemplatePool>,
        processorListRegistry: Registry<StructureProcessorList>,
        poolRL: ResourceLocation,
        nbtPieceRL: ResourceLocation,
        weight: Int,
        projection: StructureTemplatePool.Projection,
        processorListKey: ResourceKey<StructureProcessorList>,
        shouldUseLegacySingePoolElement: Boolean = false
    ) {
        if (processorListRegistry.getHolder(processorListKey).isEmpty) {
            return
        }
        val processorList = processorListRegistry.getHolder(processorListKey).get()
        val pool = templatePoolRegistry[poolRL] as? StructurePoolAccessor ?: return
        val piece = if (shouldUseLegacySingePoolElement) {
            LegacySinglePoolElement.single(nbtPieceRL.toString(), processorList).apply(projection)
        } else {
            SinglePoolElement.single(nbtPieceRL.toString(), processorList).apply(projection)
        }
        repeat(times = weight) { pool.elements.add(piece) }
        val listOfPieceEntries = ArrayList(pool.getElementCounts())
        listOfPieceEntries.add(Pair(piece, weight))
        pool.elements.add(piece)
        pool.setElementCounts(listOfPieceEntries)
    }
}

